File: kernel/mem.go
package main

import "unsafe"

type Phys uint64

type SliceHeader struct {
	Data uintptr
	Len  int
	Cap  int
}

func cleartlb()

/*
	the whole address space               is 256 TB
	a page directory pointer table addresses 512 GB
	a page directory               addresses   1 GB
	a page table                   addresses   2 MB
	a page                         addresses   4 KB
*/

const (
	COREMAPSIZE   = 512
	max32         = 0xFFFFFFFF
	max64         = 0xFFFFFFFFFFFFFFFF
	kernelstart   uintptr = 0x100000
	E820FREE      uint64  = 1
	E820RSVD      uint64  = 2
	MAXE820       = 100   // maximum number of E820 entries
	PAGESIZE      = 4096
	LARGEPAGESIZE = 2097152
	PAGETABLESIZE = 512
	PAGEAVAIL     uint64 = 1
	PAGEWRITE     uint64 = 2
	PAGEUSER      uint64 = 4
	PAGELARGE     uint64 = 0x80
	ANTIPAGE      = max32 ^ (PAGESIZE - 1)
)

type coreMapEntry struct {
	start   uint64
	end     uint64
	virtual uintptr
}

var (
	e820map     [][3]uint64
	e820num     int
	coremap     [COREMAPSIZE]coreMapEntry
	cmsize      int
	memsize     uint64
	maxvirtaddr uintptr
	kernelpml4  []uint64
)

func pageroundup(n uint64) uint64 {
	return (n + PAGESIZE - 1) & (max64 ^ (PAGESIZE - 1))
}

func processe820() {
	var e820limits [2 * MAXE820]uint64
	k := 0
	for i := 0; i < e820num; i++ {
		e820limits[k] = pageroundup(e820map[i][0])
		e820limits[k+1] = pageroundup(e820map[i][0] + e820map[i][1])
		k += 2
	}
	swapped := true
	for swapped {
		swapped = false
		for i := 1; i < k; i++ {
			if e820limits[i] == e820limits[i-1] {
				e820limits[i], e820limits[k-1] = e820limits[k-1], e820limits[i]
				swapped = true
				k--
			}
			if e820limits[i] < e820limits[i-1] {
				e820limits[i], e820limits[i-1] = e820limits[i-1], e820limits[i]
				swapped = true
			}
		}
	}
	cmsize = 0
	memsize := uint64(0)
	for i := 0; i < k-1; i++ {
		l := e820limits[i]
		found := false
		for j := 0; j < e820num; j++ {
			if l >= e820map[j][0] && l < e820map[j][0]+e820map[j][1] {
				if e820map[j][2] != E820FREE {
					goto cont
				} else {
					found = true
				}
			}
		}
		if found {
			coremap[cmsize] = coreMapEntry{start: l, end: e820limits[i+1]}
			memsize += e820limits[i+1] - l
			cmsize++
		}
	cont:
	}
	coremap[cmsize] = coreMapEntry{}
	if memsize < 16777216 {
		fuck("Sorry, GOFY doesn't run on toasters")
	}
	print((memsize + 524288) / 1048576)
	print(" MB memory\n")
}

func mapmemory() {
	pmlo, pdpo, pdo := 0, 0, 1
	pml := (*[PAGETABLESIZE]uint64)(unsafe.Pointer(uintptr(0x1000)))
	pdp := (*[PAGETABLESIZE]uint64)(unsafe.Pointer(uintptr(0x2000)))
	pd := (*[PAGETABLESIZE]uint64)(unsafe.Pointer(uintptr(0x3000)))
	virtual := uintptr(LARGEPAGESIZE)
	physical := uint64(LARGEPAGESIZE)
	offset := uintptr(0x4000)
	k := 0
	for ; k < cmsize; k++ {
		if coremap[k].start <= physical && physical < coremap[k].end {
			goto found
		}
	}
	fuck("E820 error")
found:
	coremap[k].virtual = virtual - uintptr(physical - coremap[k].start)
	for {
		pd[pdo] = physical | PAGELARGE | PAGEWRITE | PAGEAVAIL
		pdo++
		if pdo == PAGETABLESIZE {
			pd = (*[PAGETABLESIZE]uint64)(unsafe.Pointer(offset))
			pdp[pdpo] = uint64(offset) | PAGEWRITE | PAGEAVAIL
			offset += PAGESIZE
			pdpo++
			if pdpo == PAGETABLESIZE {
				pdp = (*[PAGETABLESIZE]uint64)(unsafe.Pointer(offset))
				pml[pmlo] = uint64(offset) | PAGEWRITE | PAGEAVAIL
				offset += PAGESIZE
				pmlo++
			}
		}
		virtual += LARGEPAGESIZE
		physical += LARGEPAGESIZE
		if physical >= coremap[k].end {
		retry:
			k++
			if k >= cmsize {
				break
			}
			physical = coremap[k].start
			if physical+LARGEPAGESIZE > coremap[k].end {
				goto retry
			}
			coremap[k].virtual = virtual - uintptr(physical - coremap[k].start)
		}
	}
	maxvirtaddr = virtual
	cleartlb()
}

func initmem() {
	e820num = int(*(*uint32)(unsafe.Pointer(uintptr(0x600))))
	if e820num == 0 {
		fuck("E820 fucked up")
	}
	if e820num > MAXE820 {
		fuck("E820 map too large")
	}

	mh := (*SliceHeader)(unsafe.Pointer(&e820map))
	mh.Data = 0x608
	mh.Len = e820num
	mh.Cap = mh.Len
	processe820()
	mapmemory()
	h := (*SliceHeader)(unsafe.Pointer(&kernelpml4))
	h.Data = 0x1000
	h.Len = 512
	h.Cap = 512
}

func fuck(s string) {
	println("SHIT IS BROKEN")
	println(s)
	for {
	}
}

func DecodePhys(p Phys) uintptr {
	if p < LARGEPAGESIZE {
		return uintptr(p)
	}
	for i := 0; i < cmsize; i++ {
		if p >= Phys(coremap[i].start) && p < Phys(coremap[i].end) {
			return uintptr(p) - uintptr(coremap[i].start) + coremap[i].virtual
		}
	}
	fuck("DecodePhys: tried accessing unmapped address")
	return 0xDEADBEEFDEADBEEF
}

func DecodeVirt(v uintptr) Phys {
	if v < LARGEPAGESIZE {
		return Phys(v)
	}
	for i := 0; i < cmsize; i++ {
		if v >= coremap[i].virtual && v < uintptr(coremap[i].end-coremap[i].start)+coremap[i].virtual {
			return Phys(v) - Phys(coremap[i].virtual) + Phys(coremap[i].start)
		}
	}
	fuck("DecodeVirt: tried accessing unmapped address")
	return 0xDEADBEEFDEADBEEF
}

func MakeTableSlice(p Phys) (res []uint64) {
	h := (*SliceHeader)(unsafe.Pointer(&res))
	h.Data = DecodePhys(p)
	h.Len = 512
	h.Cap = 512
	return
}

func AddrDecode(v uintptr) (pdp, pd, pt, page, disp int) {
	pdp = int((v >> 39) & 0x1FF)
	pd = int((v >> 30) & 0x1FF)
	pt = int((v >> 21) & 0x1FF)
	page = int((v >> 12) & 0x1FF)
	disp = int(v & (PAGESIZE - 1))
	return
}

File: kernel/runtime/mem.c
#include "runtime.h"

/*
	the whole address space               is 256 TB
	a page directory pointer table addresses 512 GB
	a page directory               addresses   1 GB
	a page table                   addresses   2 MB
	a page                         addresses   4 KB
*/

enum {
	COREMAPSIZE   = 512,
	KERNELSTART   = 0x100000,
	E820FREE      = 1,
	E820RSVD      = 2,
	MAXE820       = 100,   // maximum number of E820 entries
	PAGESIZE      = 4096,
	LARGEPAGESIZE = 2097152,
	PAGETABLESIZE = 512,
	PAGEAVAIL     = 1,
	PAGEWRITE     = 2,
	PAGEUSER      = 4,
	PAGELARGE     = 0x80,
	ANTIPAGE      = ~(PAGESIZE - 1),
	TMPPAGESTART  = 0x3fe00000LL
};

void main·fuck(int8*, uint32);
void runtime·FlushTLB(void);
void runtime·SetCR3(uint64*);
void runtime·InvlPG(void*);

typedef struct coreMapEntry coreMapEntry;

struct coreMapEntry {
	uint64 start, end;
};

uint64* e820map;
uint32 e820num;
coreMapEntry coremap[COREMAPSIZE];
uint32 cmsize;
uint64 memsize;
uint64 e820limits[2 * MAXE820];
uint64 runtime·highest;
extern uint64 *runtime·TmpPageTable;
uint64 *pml4;
extern uint64 runtime·KernelPD;

uint64
pageroundup(uint64 n)
{
	return (n + PAGESIZE - 1) & ANTIPAGE;
}

#pragma textflag 7
uint64
falloc(uint64 n)
{
	uint32 i;
	uint64 p;

	if(!n) return 0;
        for(i = 0; i < cmsize; i++) {
                if(coremap[i].end - coremap[i].start >= PAGESIZE*n) {
                        p = coremap[i].start;
                        coremap[i].start += PAGESIZE * n;
                        if(coremap[i].start == coremap[i].end) {
                                i++;
				cmsize--;
                                for(; i < cmsize; i++) coremap[i-1] = coremap[i];
                        }
                        return p;
                }
        }
	static int8 s[] = "out of memory";
	main·fuck(s, sizeof(s));
        return 0;
}

void
runtime·Falloc(uint64 n, uint64 r)
{
	r = falloc(n);
	FLUSH(&r);
}

#pragma textflag 7
void
ffree(uint64 p, uint64 n)
{
	uint32 i;

	if(!n) return;
	for(i = 0; i < cmsize && coremap[i].start <= i; i++);
	if(i && coremap[i-1].end == p) {
		coremap[i-1].end += n;
		if(p+n == coremap[i].start) {
			coremap[i-1].end = coremap[i].end;
			i++;
			cmsize--;
			for(; i < cmsize; i++) {
				coremap[i-1] = coremap[i];
			}
		}
	} else {
		if(p+n == coremap[i].start) {
			coremap[i].start -= n;
		} else {
			coreMapEntry *j;

			j = coremap + i;
			for(; i < cmsize; i++) {
				coremap[i+1] = coremap[i];
			}
			cmsize++;
			j->start = p;
			j->end = p+n;
		}
        }
}

void
runtime·Ffree(uint64 n, uint64 r)
{
	ffree(n, r);
}


#pragma textflag 7
void
runtime·processe820(void)
{
	uint32 i, j;
	uint64 *l, *lk, t;
	bool swapped, found;

	lk = e820limits;
	for(i = 0; i < e820num; i++) {
		e820map[3*i] = pageroundup(e820map[3*i]);
		e820map[3*i+1] &= ANTIPAGE;
		*lk++ = e820map[3*i];
		*lk++ = e820map[3*i] + e820map[3*i+1];
	}
	swapped = true;
	while(swapped) {
		swapped = false;
		for(l=e820limits + 1; l < lk; l++) {
			if(*l == *(l-1)) {
				lk--;
				t = *l;
				*l = *lk;
				*lk = t;
				swapped = true;
			}
			if(*l < *(l-1)) {
				t = *l;
				*l = *(l-1);
				*(l-1) = t;
				swapped = true;
			}
		}
	}
	cmsize = 0;
	memsize = 0;
	for(l = e820limits; l < lk-1; l++) {
		found = false;
		for(j = 0; j < e820num; j++) {
			if(*l >= e820map[3*j] && *l < e820map[3*j]+e820map[3*j+1]) {
				if(e820map[3*j+2] != E820FREE) {
					goto cont;
				} else {
					found = true;
				}
			}
		}
		if(found) {
			if(cmsize && coremap[cmsize-1].end == *l) {
				coremap[cmsize-1].end = *(l+1);
				memsize += *(l+1) - *l;
			} else {
				coremap[cmsize].start = *l;
				coremap[cmsize].end = *(l+1);
				memsize += *(l+1) - *l;
				cmsize++;
			}
		} 
	cont: ;
	}
	/* this is considered "reserved" even though being perfectly usable memory -- it's just already allocated */
	memsize += runtime·highest - KERNELSTART;
	coremap[cmsize].start = 0;
	coremap[cmsize].end = 0;
	runtime·printf("%d MB core\n", (uint32)((memsize + 524288) / 1048576));
	if(memsize < 16777216) {
		static int8 s[] = "Sorry, GOFY doesn't run on toasters";
		main·fuck(s, sizeof(s));
	}
}

#pragma textflag 7
void
runtime·mapmemory(void)
{
	uint64 addr;
	uint32 pmlo, pdpo, pdo, pto;
	uint64 *pdp, *pd, *pt;

	pml4 = (uint64*) falloc(1);
	runtime·memclr((uint8*) pml4, PAGESIZE);
	pmlo = pdpo = pdo = pto = 511;
	pdp = pd = pt = 0; // make the compiler happy
	for(addr=0; addr < runtime·highest; addr += 4096) {
		pto++;
		if(pto == 512) {
			pdo++;
			pto = 0;
			if(pdo == 512) {
				pdpo++;
				pdo = 0;
				if(pdpo == 512) {
					pmlo++;
					if(pmlo == 512) pmlo = 0;
					pdpo = 0;
					pdp = (uint64*) falloc(1);
					pml4[pmlo] = ((uint64) pdp) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
					runtime·memclr((uint8*) pdp, PAGESIZE);
				}
				pd = (uint64*) falloc(1);
				pdp[pdpo] = ((uint64) pd) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
				runtime·memclr((uint8*) pd, PAGESIZE);
			}
			pt = (uint64*) falloc(1);
			pd[pdo] = ((uint64) pt) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
			runtime·memclr((uint8*) pt, PAGESIZE);
		}
		pt[pto] = addr | PAGEAVAIL | PAGEWRITE;
	}
	pdp = (uint64*) (pml4[0] & ANTIPAGE);
	runtime·KernelPD = pdp[0] & ANTIPAGE;
	pd = (uint64*) runtime·KernelPD;
	runtime·TmpPageTable = (uint64*) falloc(1);
	runtime·memclr((uint8*) runtime·TmpPageTable, PAGESIZE);
	pd[511] = ((uint64) runtime·TmpPageTable) | PAGEAVAIL | PAGEWRITE;

	pt = (uint64*) (pd[0] & ANTIPAGE);
	pt[0] = 0;
	
	runtime·SetCR3(pml4);
}

#pragma textflag 7
void
runtime·SysMemInit(void)
{
	e820num = *(uint32*)0x600;
	if(e820num == 0) {
		static int8 s[] = "E820 fucked up";
		main·fuck(s, sizeof(s));
	}
	if(e820num > MAXE820) {
		static int8 s[] = "E820 map too large";
		main·fuck(s, sizeof(s));
	}
	runtime·highest = pageroundup(runtime·highest);
	e820map = (uint64*) 0x608;
	e820map[3*e820num+0] = KERNELSTART;
	e820map[3*e820num+1] = runtime·highest - KERNELSTART;
	e820map[3*e820num+2] = E820RSVD;
	e820map[3*e820num+3] = 0;
	e820map[3*e820num+4] = 0x10000;
	e820map[3*e820num+5] = E820RSVD;
	e820num += 2;
	runtime·processe820();
	runtime·mapmemory();
}

uint32 tmpref[512];

#pragma textflag 7
void*
maptmp(uint64 phys)
{
	void *r;
	uint32 i;
	if((phys & (PAGESIZE - 1)) != 0) {
		static int8 s[] = "MapTmp called with invalid address";
		main·fuck(s, sizeof(s));
	}

	for(i=0;i<512;i++) {
		if((runtime·TmpPageTable[i] & ANTIPAGE) == phys) {
			tmpref[i]++;
			return (void*) (TMPPAGESTART + i * PAGESIZE);
		}
	}
	for(i=0;i<512;i++) {
		if(tmpref[i] == 0) {
			tmpref[i]++;
			runtime·TmpPageTable[i] = phys | PAGEAVAIL | PAGEWRITE;
			r = (void*) (TMPPAGESTART + i * PAGESIZE);
			runtime·InvlPG(r);
			return r;
		}
	}
	static int8 s[] = "out of temporary pages";
	main·fuck(s, sizeof(s));
	return 0;
}

void
runtime·MapTmp(uint64 phys, void* r)
{
	r = maptmp(phys);
	FLUSH(&r);
}

#pragma textflag 7
void
runtime·FreeTmp(void* t)
{
	uint32 i;

	if(t == 0) return;
	if(t < (void*) TMPPAGESTART) {
		static int8 s[] = "FreeTmp called with invalid address";
		main·fuck(s, sizeof(s));
	}
	i = ((uint64)t - TMPPAGESTART) / PAGESIZE;
	if(tmpref[i] > 0) tmpref[i]--;
}

#pragma textflag 7
void
runtime·MapMem(uint64 pmlphys, uint64 phys, void* virt, uint32 n, uint32 flags)
{
	uint32 pmlo, pdpo, pdo, pto;
	uint64 *pml, *pdp, *pd, *pt;

	if((uint64)phys & ~ANTIPAGE || (uint64)virt & ~ANTIPAGE) {
		static int8 s[] = "MapMem called with invalid address";
		main·fuck(s, sizeof(s));
	}

	pml = maptmp(pmlphys);
	pmlo = ((uint64)virt >> 39) & 0x1FF;
	pdpo = ((uint64)virt >> 30) & 0x1FF;
	pdo = ((uint64)virt >> 21) & 0x1FF;
	pto = ((uint64)virt >> 12) & 0x1FF;
	if((pml[pmlo] & PAGEAVAIL) == 0) {
		pml[pmlo] = falloc(1) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
		pdp = maptmp(pml[pmlo] & ANTIPAGE);
		runtime·memclr((uint8*) pdp, PAGESIZE);
	} else {
		pdp = maptmp(pml[pmlo] & ANTIPAGE);
	}
	if((pdp[pdpo] & PAGEAVAIL) == 0) {
		pdp[pdpo] = falloc(1) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
		pd = maptmp(pdp[pdpo] & ANTIPAGE);
		runtime·memclr((uint8*) pd, PAGESIZE);
	} else {
		pd = maptmp(pdp[pdpo] & ANTIPAGE);
	}
	if((pd[pdo] & PAGEAVAIL) == 0) {
		pd[pdo] = falloc(1) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
		pt = maptmp(pd[pdo] & ANTIPAGE);
		runtime·memclr((uint8*) pt, PAGESIZE);
	} else {
		pt = maptmp(pd[pdo] & ANTIPAGE);
	}

	while(n--) {
		pt[pto] = phys | PAGEAVAIL | PAGEWRITE | flags;
		pto++;
		if(pto == PAGETABLESIZE) {
			pto = 0;
			pdo++;
			runtime·FreeTmp(pt);
			if(pdo == PAGETABLESIZE) {
				pdo = 0;
				pdpo++;
				runtime·FreeTmp(pd);
				if(pdpo == PAGETABLESIZE) {
					pdpo = 0;
					pmlo++;
					runtime·FreeTmp(pdp);
					if((pml[pmlo] & PAGEAVAIL) == 0) {
						pml[pmlo] = falloc(1) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
						pdp = maptmp(pml[pmlo] & ANTIPAGE);
						runtime·memclr((uint8*) pdp, PAGESIZE);
					} else {
						pdp = maptmp(pml[pmlo] & ANTIPAGE);
					}
				}
				if((pdp[pdpo] & PAGEAVAIL) == 0) {
					pdp[pdpo] = falloc(1) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
					pd = maptmp(pdp[pdpo] & ANTIPAGE);
					runtime·memclr((uint8*) pd, PAGESIZE);
				} else {
					pd = maptmp(pdp[pdpo] & ANTIPAGE);
				}
			}
			if((pd[pdo] & PAGEAVAIL) == 0) {
				pd[pdo] = falloc(1) | PAGEAVAIL | PAGEWRITE | PAGEUSER;
				pt = maptmp(pd[pdo] & ANTIPAGE);
				runtime·memclr((uint8*) pt, PAGESIZE);
			} else {
				pt = maptmp(pd[pdo] & ANTIPAGE);
			}
		}
		phys += PAGESIZE;
	}
	runtime·FreeTmp(pml);
	runtime·FreeTmp(pdp);
	runtime·FreeTmp(pd);
	runtime·FreeTmp(pt);
}

#pragma textflag 7
void
runtime·SysFree()
{
}

#pragma textflag 7
void*
runtime·SysAlloc(uintptr n)
{
	uint64 phys;
	void* virt;

	virt = (void*) runtime·highest;
	n = (n + PAGESIZE - 1) / PAGESIZE;
	phys = falloc(n);
	runtime·highest += n * PAGESIZE;
	runtime·MapMem((uint64) pml4, phys, virt, n, 0);
	runtime·FlushTLB();
	runtime·memclr(virt, n * PAGESIZE);
	return virt;
}

#pragma textflag 7
void
runtime·AlignAlloc(uintptr n, void* virt, uintptr phys)
{
	virt = (void*) runtime·highest;
	n = (n + PAGESIZE - 1) / PAGESIZE;
	phys = falloc(n);
	runtime·highest += n * PAGESIZE;
	runtime·MapMem((uint64) pml4, phys, virt, n, 0);
	runtime·FlushTLB();
	runtime·memclr(virt, n * PAGESIZE);
	FLUSH(&virt);
	FLUSH(&phys);
}
